Use vcpkg on all presets

tophyr opened the pull request:
## Pull Request Type
<!-- Please select which type of change this most aligns with. If more than one type fits, please select multiple. -->

- [ ] GitHub Workflow changes
- [ ] Documentation or Wiki changes
- [x] Build and Dependency changes
- [ ] Runtime changes
  - [ ] Render changes
  - [ ] Audio changes
  - [ ] Input changes
  - [ ] Network changes
  - [ ] Other changes

### Description
Uses vcpkg for all platforms and removes platform-specific installation methods for ZLib, SDL and GTest. Still allows system-provided libraries.

### Related Issues
Effectively a rebased and simplified reboot of @Arcnor's #242

### Checklist
<!-- Please review the following checklist before submitting your pull request -->

- [x] I have tested my changes locally and verified that they work as intended.
- [x] I have documented any new or modified functionality.
- [x] I have reviewed the changes to ensure they do not introduce any unnecessary complexity or duplicate code.
- [x] I understand that by submitting this pull request, I am agreeing to license my contributions under the project's license.

Lgt2x commented on .github/workflows/build.yml:
```diff
@@ -66,22 +66,25 @@ jobs:
           sudo apt update
           sudo apt install -y --no-install-recommends \
-            ninja-build cmake g++ libgtest-dev libsdl2-dev zlib1g-dev libspdlog-dev
+            ninja-build cmake g++ libspdlog-dev
```
you can probably also remove spdlog, the dependency has been dropped in #341

tophyr replied:
yeah, i noticed that - am fine to do, just figured i'd keep this to a minimal change

Lgt2x replied:
While we're editing the file, we can add that change ;)

Lgt2x commented on CMakePresets.json:
```diff
@@ -39,8 +21,7 @@
 				"strategy": "external",
 				"value": "x64"
-			},
-			"toolchainFile": "$env{VCPKG_ROOT}/scripts/buildsystems/vcpkg.cmake"
```
I think you still need that to load the toolchain, especially locally

Lgt2x replied:
Is there a reason we don't set the toolchain in the presets? I see that now `VCPKG_TOOLCHAIN_FILE` is set by us in CMakeLists, can't we keep it in the toolchain anyways?

tophyr commented:
I added it into the main CMakeLists right at the top. That accomplishes two
things: it removes the burden of specifying it on the invocation (or on
specifying a preset) and it lets us tie in to vcpkg's toolchain-chaining
capability, which will be critical for Android.

On Wed, Jun 26, 2024 at 12:51 PM Louis Gombert ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In CMakePresets.json
> <https://github.com/DescentDevelopers/Descent3/pull/469#discussion_r1655297212>
> :
>
> > @@ -39,8 +21,7 @@
>  			"architecture": {
>  				"strategy": "external",
>  				"value": "x64"
> -			},
> -			"toolchainFile": "$env{VCPKG_ROOT}/scripts/buildsystems/vcpkg.cmake"
>
> I think you still need that to load the toolchain, especially locally
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/DescentDevelopers/Descent3/pull/469#pullrequestreview-2142585308>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAPB6RKC76GU7MYXKHKS5RLZJL5S7AVCNFSM6AAAAABJ5B7LOKVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDCNBSGU4DKMZQHA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

Lgt2x commented:
> I added it into the main CMakeLists right at the top. That accomplishes two things: it removes the burden of specifying it on the invocation (or on specifying a preset) and it lets us tie in to vcpkg's toolchain-chaining capability, which will be critical for Android.
> […](#)
> On Wed, Jun 26, 2024 at 12:51 PM Louis Gombert ***@***.***> wrote: ***@***.**** commented on this pull request. ------------------------------ In CMakePresets.json <[#469 (comment)](https://github.com/DescentDevelopers/Descent3/pull/469#discussion_r1655297212)> : > @@ -39,8 +21,7 @@ "architecture": { "strategy": "external", "value": "x64" - }, - "toolchainFile": "$env{VCPKG_ROOT}/scripts/buildsystems/vcpkg.cmake" I think you still need that to load the toolchain, especially locally — Reply to this email directly, view it on GitHub <[#469 (review)](https://github.com/DescentDevelopers/Descent3/pull/469#pullrequestreview-2142585308)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAPB6RKC76GU7MYXKHKS5RLZJL5S7AVCNFSM6AAAAABJ5B7LOKVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDCNBSGU4DKMZQHA> . You are receiving this because you authored the thread.Message ID: ***@***.***>

Oh right, I did not notice, that's fine

Jayman2000 commented on README.md:
```diff
@@ -92,6 +92,10 @@ The milestone needs testing on all platforms. Please report issues when found.
 
 ## Building
+#### Prerequisites
+1. [**vcpkg**](https://github.com/microsoft/vcpkg)
+    This project depends on [`vcpkg`](https://github.com/microsoft/vcpkg), a cross-platform dependency-management system developed by Microsoft. You will need to install it locally by cloning https://github.com/microsoft/vcpkg and setting your `VCPKG_ROOT` env var to the repository location.
+
```
This kind of breaks the order of the build instructions. In order to clone the `vcpkg` repo, you need to either have the `git` command or something like [GitHub Desktop](https://desktop.github.com), but we don’t tell users to install Git until later on in the instructions.

Additionally, this step is redundant on Windows. [When you open an “x86 Native Tools Command Prompt”, it will set `VCPKG_ROOT` for you.](https://github.com/DescentDevelopers/Descent3/pull/413#discussion_r1624974614)

Also, I have a few minor critiques of this part:

- Having a list with a single item on it looks weird. I would remove the `1. `.
- The bold `vcpkg` link is redundant. I would remove the `[**vcpkg**](https://github.com/microsoft/vcpkg)`.

tophyr replied:
Updated! I'm thinking that, in a separate PR, I'll offer up a way to make `vcpkg` a submodule of the main git repo - that would eliminate the need for a prereq setup step in the first place. For this PR however I think just the doc update should be sufficient; don't wanna scope-creep.

Jayman2000 replied:
OK. I think that it’s good enough at the moment. We can wait until the submodule PR to fix the build instructions order.

(Again, I would mark this conversation as resolved, but I can’t.)

tophyr commented on .github/workflows/build.yml:
```diff
@@ -66,22 +66,25 @@ jobs:
 
+      - uses: lukka/run-vcpkg@v11
+        with:
+          vcpkgGitCommitId: 2e7fd315d713e9eac52f6bdb24afbf1b4dbbe1e5
```
awesome! will update. didn't know the difference there

tophyr replied:
updated.

Jayman2000 replied:
Sweet! I would mark this conversation as resolved, but [I can’t because I don’t have write access to this repo](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#resolving-conversations).

tophyr commented on CMakeLists.txt:
```diff
@@ -1,5 +1,15 @@
 cmake_minimum_required(VERSION 3.20)  # For using CMAKE_<LANG>_BYTE_ORDER
 
+if (NOT DEFINED ENV{VCPKG_ROOT})
+  message(FATAL_ERROR "
+    Please set an environment variable VCPKG_ROOT
+    For example:
+    export VCPKG_ROOT=/path/to/vcpkg
+    ")
+endif()
```
interesting thought. in some sense, yes - i think you would not be able to _build_ using a system-provided library.

HOWEVER, as i understand it, SDL and zlib are both dynamically linked - correct? if that's the case, then you could still _run_ Descent3 using the system-provided SDL and zlib, by having the binary load up the system libs instead of the vcpkg ones. all that is really needed for the _build_ are the headers and symbol stubs, which I do think we should keep strictly specified.

tophyr replied:
I was incorrect here: We are currently linking SDL statically.

winterheart replied:
This adds additional dependency vcpkg for users on Linux and macOS, which on some cases undesirable (Linux distribution enforces using external dependencies from thier repositories for security reasons). We should keep ability to install external dependencies without using vcpkg.

winterheart commented on Brewfile:
```diff
@@ -1,11 +1,4 @@
 # Homebrew dependencies to build Descent3
 
-# SDL
-brew "sdl2"
-
 brew "cmake"
-brew "googletest"
 brew "ninja"
-
-# zlib
-brew "zlib"
```
We should preserve ability to install external dependencies without using vcpkg for users. Brewfile should left unchanged. You can override this in CI with direct brew invocation, i.e. `brew install cmake ninja`.

tophyr replied:
While I'm all for a solution that enables using system-provided libraries, I'm not sure I agree that the Brewfile we provide should automatically install the build deps - if the user wants to install those deps via Brew then more power to them, but I think if our scripts install them then they should come from vcpkg.

winterheart reviewed (commented):
This adds additional required dependency vcpkg for users on Linux and macOS, which on some cases undesirable (Linux distribution enforces using external dependencies from thier repositories for security reasons). We should keep ability to install external dependencies without using vcpkg.

tophyr commented:
@winterheart I'd be happy with a solution that accepted either system-installed libraries or vcpkg ones. I would assume that a system-packaged library should take precedence over a vcpkg-provided one? Is there a pre-existing convention to use for buildscript parameters (`-DCMAKE_FOO` etc) to force using one or the other? (For example, in the case that I do have a system SDL installed but I want the build to use the vcpkg one anyway.)

Jayman2000 commented:
> @winterheart I'd be happy with a solution that accepted either system-installed libraries or vcpkg ones.

[What about this?](https://github.com/Jayman2000/Descent3-pr/commit/5e0bcc74d65fd8f67b974c0a82c3cb35b3407e47)

> I would assume that a system-packaged library should take precedence over a vcpkg-provided one?

I don’t think that the system-packaged library should take precedence over a vcpkg-provided one. I want to be able to easily test two scenarios:

1. Compiling Descent 3 with all dependencies provided by vcpkg. Testing this scenario helps ensure that there are no undeclared dependencies.
2. Compiling Descent 3 with all dependencies provided by the system package manager. Testing this scenario helps ensure that packages for Descent 3 be built in a way that distributions will approve of.

If system packages took precedence over vcpkg packages, then it would be more difficult to test that first scenario.

> Is there a pre-existing convention to use for buildscript parameters (`-DCMAKE_FOO` etc) to force using one or the other? (For example, in the case that I do have a system SDL installed but I want the build to use the vcpkg one anyway.)

I don’t know if there’s a pre-existing convention, but I do know that [GZDoom has options like `FORCE_INTERNAL_BZIP2` that you might want to take a look at.](https://github.com/ZDoom/gzdoom/blob/ddbf90389b34c926aee11dea92f8cf0787198f74/CMakeLists.txt#L335)

winterheart commented:
> @winterheart I'd be happy with a solution that accepted either system-installed libraries or vcpkg ones. I would assume that a system-packaged library should take precedence over a vcpkg-provided one? Is there a pre-existing convention to use for buildscript parameters (`-DCMAKE_FOO` etc) to force using one or the other? (For example, in the case that I do have a system SDL installed but I want the build to use the vcpkg one anyway.)

I think -DUSE_VCPKG (with default OFF) option would be OK. I don't think, that we'll need to install different dependencies from different sources.

Lgt2x commented:
Why not try to use VCPKG if available (ie VCPKG_ROOT env variable set) and fallback to system libs if not found? With a clear message during configuration that tells you what's being used. I agree though that mixing VCPKG and systems libs is a bad idea

tophyr commented:
Went with a combination of @Jayman2000's and @Lgt2x's ideas: The system will use vcpkg by default if `VCPKG_ROOT` is set in the environment, _and_ if `USE_VCPKG` is not set to `OFF`.

Something I've noticed is that if the build has already been configured, it will continue using that configuration's settings even if different ones are specified. Is that normal CMake behavior? This feels like either a bug or a misunderstanding on my part.

Lgt2x commented on CMakeLists.txt:
```diff
@@ -1,5 +1,19 @@
+option(USE_VCPKG "
+  When the VCPKG_ROOT environment is set, the Descent3 build will source its dependencies from VCPKG.
+  Set USE_VCPKG=OFF, or unset VCPKG_ROOT, to source dependencies from the system libraries.
+  If the build was previously configured to the opposite, you will need to delete the build dir first.
```
that shouldnt be the case

Lgt2x commented:
> Went with a combination of @Jayman2000's and @Lgt2x's ideas: The system will use vcpkg by default if `VCPKG_ROOT` is set in the environment, _and_ if `USE_VCPKG` is not set to `OFF`.
> 
> Something I've noticed is that if the build has already been configured, it will continue using that configuration's settings even if different ones are specified. Is that normal CMake behavior? This feels like either a bug or a misunderstanding on my part.

Can you give an example of the commands you've tried?

Needs a warning or error message when USE_VCPKG is ON and VCPKG_ROOT cannot be found

tophyr commented:
> Can you give an example of the commands you've tried?

It appears to be related to the `CMAKE_TOOLCHAIN_FILE` persisting, not the other options: https://gist.github.com/tophyr/2f8df17c6a46c3a9157ac73f69de1939

tophyr commented:
Hm, looks to be related to CMakeCache.txt

tophyr commented:
Rewrote the logic to enable warning if `USE_VCPKG && !VCPKG_ROOT` per @Lgt2x's feedback. Behavior is now:

| USE_VCPKG | ENV{VCPKG_ROOT} | Behavior |
| -- | -- | -- |
| \<unset\> | \<unset\> | System Libraries |
| \<unset\> | \<set\> | vcpkg |
| `ON` | \<unset\> | Warning + System Libraries |
| `ON` | \<set\> | vcpkg |
| `OFF` | \<unset\> | System Libraries |
| `OFF` | \<set\> | System Libraries |

Rewrote the Build Instructions in the README file as well; feedback welcome.

Lgt2x commented on README.md:
```diff
@@ -114,61 +121,130 @@ The milestone needs testing on all platforms. Please report issues when found.
+
+| Option | Description | Default |
+| ------ | ----------- | ------- |
+| `--config` | `Debug` builds are generally larger, slower and contain extra correctness checks that will validate game data and interrupt gameplay when problems are detected. `Release` builds are optimized for size and speed and do not include debugging information, which makes it harder to find problems. | `Debug` |
```
You can say that using `CMAKE_BUILD_TYPE` is the same

Lgt2x commented on CMakeLists.txt:
```diff
@@ -1,5 +1,25 @@
 cmake_minimum_required(VERSION 3.20)  # For using CMAKE_<LANG>_BYTE_ORDER
 
+if(NOT DEFINED USE_VCPKG OR USE_VCPKG)
```
`USE_VCPKG` should always be defined if you create the corresponding `option(`

tophyr replied:
> While we're editing the file, we can add that change and remove spdlog ;)

done

> `USE_VCPKG` should always be defined if you create the corresponding `option()`

With the desired behavior, `USE_VCPKG` needs to be a tristate: explicitly true (so we can warn if vcpkg is not found), explicitly false (so we can ignore vcpkg if it is set), or undefined (so we can go by existence of vcpkg). Thus, I don't declare it as an `option()` - I just document it in the README.

Lgt2x replied:
If a variable should be set by the user, then it should be an option or cache variable. If it has more than 2 state, then it should be set as string cache variable, ie `USE_VCPKG` as "AUTOMATIC", "FORCE_VCPKG" or "FORCE_SYSTEM".

tophyr replied:
how should that be expressed in the cmake file?

Lgt2x replied:
Similarly to the example with CRYPTOBACKEND in this doc page https://cmake.org/cmake/help/book/mastering-cmake/chapter/CMake%20Cache.html

tophyr replied:
> Similarly to the example with CRYPTOBACKEND in this doc page https://cmake.org/cmake/help/book/mastering-cmake/chapter/CMake%20Cache.html

Feedback incorporated in https://github.com/DescentDevelopers/Descent3/pull/469/commits/83b21c4a8021a9e16a017cbdc55fb80abafee31b

Lgt2x commented on CMakeLists.txt:
```diff
@@ -1,5 +1,25 @@
 cmake_minimum_required(VERSION 3.20)  # For using CMAKE_<LANG>_BYTE_ORDER
 
+if(NOT DEFINED USE_VCPKG OR USE_VCPKG)
+  message("NOT DEFINED USE_VCPKG OR USE_VCPKG")
```
These look like internal debug messages. User output should be `Using system-provided libraries` or `Using VCPKG` if you want to keep some feedback

tophyr replied:
🤦‍♂️

Lgt2x commented on CMakeLists.txt:
```diff
@@ -1,5 +1,25 @@
+  message("NOT DEFINED USE_VCPKG OR USE_VCPKG")
+  if(DEFINED ENV{VCPKG_ROOT})
+    message("DEFINED ENV{VCPKG_ROOT}")
+    if (CMAKE_TOOLCHAIN_FILE)
```
When is this variable set? Only for cross-compilation builds?

tophyr replied:
see https://github.com/DescentDevelopers/Descent3/pull/469#discussion_r1673465920

Lgt2x commented on CMakeLists.txt:
```diff
@@ -1,5 +1,25 @@
+    cmake_path(ABSOLUTE_PATH VCPKG_TOOLCHAIN_FILE BASE_DIRECTORY $ENV{VCPKG_ROOT} NORMALIZE OUTPUT_VARIABLE CMAKE_TOOLCHAIN_FILE)
+
+    if(CMAKE_TOOLCHAIN_FILE STREQUAL VCPKG_CHAINLOAD_TOOLCHAIN_FILE)
+      # prevent endless recursion
```
How can recursion happen?

tophyr replied:
`VCPKG_CHAINLOAD_TOOLCHAIN_FILE`, and the added build logic to conditionally set that variable, is ultimately the cause of the recursion problem. 

vcpkg itself requires the use of `CMAKE_TOOLCHAIN_FILE`, which effectively precludes any _other_ use of that variable. In order to "play nice" with other toolchain-based builds, vcpkg provides `VCPKG_CHAINLOAD_TOOLCHAIN_FILE` in order to actually use a different toolchain (ie, for Android or an embedded system).

Gradle (Android's build system) automatically - _and hard-codedly_ - sets `CMAKE_TOOLCHAIN_FILE` on the CMake invocation it kicks off for the C++ compilation side of its overall build. This causes a conflict: `CMAKE_TOOLCHAIN_FILE` will _always_ point to the Android NDK when building for Android, but vcpkg _requires_ that it point to vcpkg.

The solution I landed upon was to have our build file _re_-set the `CMAKE_TOOLCHAIN_FILE` when using vcpkg, copying the passed-in value into the `VCPKG_CHAINLOAD_TOOLCHAIN_FILE` variable vcpkg uses to play nice. All of this happens appropriately considering whether or not we're supposed to be using vcpkg in the first place.

---
**Now, here's where it gets fun:**

`CMakePresets.txt` causes some of these values to be set even when they're not specified on the invocation command line. 

Doing this, with `$ENV{VCPKG_ROOT}` set:
```
cmake --preset mac
```
ends up storing `USE_VCPKG=ON` and `CMAKE_TOOLCHAIN_FILE=/Users/csarbora/Projects/vcpkg/` (for me) in the CMakeCache.txt file, and then those values are already set (as if they were passed in on the command line) for subsequent CMake runs. This is intentional by CMake, so that this works:
```
cmake -B builds/mac -DUSE_VCPKG=OFF -DSOME_OTHER_OPTION=FOO
cmake --build builds/mac 
#            ^^          USE_VCPKG and SOME_OTHER_OPTION will be set even tho they are unspecified
```
When `CMAKE_TOOLCHAIN_FILE` is _passed in_ pointing to vcpkg (because the value is cached from a prior vcpkg-using configuration), `VCPKG_CHAINLOAD_TOOLCHAIN_FILE` gets set to the same value. This would end up causing vcpkg trying to chainload itself, resulting in effectively an infinite recursion. So in order to halt that, I check if the two variables are equal to each other and unset the latter if they are.

Lgt2x commented on README.md:
```diff
@@ -92,17 +92,24 @@ The milestone needs testing on all platforms. Please report issues when found.
 
+### Installing and using vcpkg
+* When building for Windows, vcpkg is already installed and configured when using the Developer Tools Command Prompt.
+* For Android, Linux or Mac (or non-standard-Windows) configurations, you will need to install vcpkg locally by cloning https://github.com/microsoft/vcpkg and setting your `VCPKG_ROOT` env var to the repository location. With this environment variable set, the build will automatically locate its dependencies.
```
Android is not technically supported yet

tophyr replied:
> Is there a reason we don't set the toolchain in the presets?

This is an old comment, but to close the loop on it: Setting it unconditionally would make all builds unconditionally use vcpkg, which was discovered to be an anti-goal.

> Android is not technically supported yet

_yet_

Jayman2000 replied:
I don’t think that we should mention Android here until we actually support Android. It would make it look like the `main` branch supports Android even though it doesn’t support Android. Also, there’s no guarantee that the `main` branch will support Android. [Your other PR](https://github.com/DescentDevelopers/Descent3/pull/459) might get accepted or it might get rejected. We might even decide that the upstream version of Descent 3 should never support Android and that Android support should be done in a fork. I don’t think that any of those scenarios are likely, but they are possible.

Lgt2x reviewed (commented):
Nice job, I'm still a little confused by the CMake logic (I know it's not been easy), but great improvement overall.

@Jayman2000 you may also want to take a look at the PR :)

Jayman2000 reviewed (changes requested):
Here’s some general feedback:

- The build instructions worked on Windows 10 64-bit. They also worked on Ubuntu 24.04, but only if I got the dependencies from the system package manager.

- The build instructions don’t work if you’re trying to build on Linux with vcpkg. I tried them on Ubuntu 24.04 and Fedora 40, and both of them gave me this error:

    ```
    CMake Error at CMakeLists.txt:98 (find_package):
      By not providing "FindSDL2.cmake" in CMAKE_MODULE_PATH this project has
      asked CMake to find a package configuration file provided by "SDL2", but
      CMake did not find one.

      Could not find a package configuration file provided by "SDL2" with any of
      the following names:

        SDL2Config.cmake
        sdl2-config.cmake

      Add the installation prefix of "SDL2" to CMAKE_PREFIX_PATH or set
      "SDL2_DIR" to a directory containing one of the above files.  If "SDL2"
      provides a separate development package or SDK, be sure it has been
      installed.
    ```

    [Here’s a full log.](https://gist.github.com/Jayman2000/2415b96c48d69314ab072d88e2f5857a)

- At the moment, the steps for each of the build instructions are in this order:

    1. Install the prerequisite build tools.

    2. Clone the Descent3 source code.

    3. Acquire the library dependencies.

    4. Build Descent3.

    With the above order, users do a task that doesn’t require the Descent3 repo, then do a task that does require the Descent3 repo, then do a task that doesn’t require the Descent3 repo, and then do a task that does require the Descent3 repo.

    I think that it would make more sense to order the instructions like this:

    1. Install the prerequisite build tools.

    2. Acquire the library dependencies.

    3. Clone the Descent3 source code.

    4. Build Descent3.

    With this new order, users do two tasks that don’t require the Descent3 repo, then users do two tasks that do require the Descent3 repo.

    It just seems more natural to use the Descent3 repo right after you clone it.

tophyr commented on README.md:
````diff
@@ -114,61 +121,130 @@ The milestone needs testing on all platforms. Please report issues when found.
+        ```sh
+        git clone https://github.com/microsoft/vcpkg
+        cd vcpkg
+        export VCPKG_ROOT=$(pwd)
````
interesting, i didn't actually know `$PWD` was a variable we could use. i habitually use `$()` syntax but i'm fine with this.

tophyr commented on README.md:
````diff
@@ -114,61 +121,130 @@ The milestone needs testing on all platforms. Please report issues when found.
+    ```
+
+3. **Acquire the library dependencies.**
+
+    * If you would like to use vcpkg:
+        ```sh
+        git clone https://github.com/microsoft/vcpkg
+        cd vcpkg
+        export VCPKG_ROOT=$(pwd)
+        # IMPORTANT: You will need $VCPKG_ROOT defined for all build runs. It is a good idea to set this in your .bashrc or equivalent.
+        ```
+    * If you would like to manage the code dependencies yourself:
+        ```sh
+        brew install sdl2 zlib googletest
+        ```
+
+4. **Build Descent3.**
+
+    ```sh
     cd Descent3
     brew bundle install
-    cmake --preset mac -D ENABLE_LOGGER=[ON|OFF]
+    cmake --preset mac
     cmake --build --preset mac --config [Debug|Release]
     ```
````
I removed a bunch of the Terminal text in general, lmk if it reads ok.

Jayman2000 replied:
Looks good to me.

tophyr commented:
All @Jayman2000 feedback incorporated, and build instructions for both Ubuntu and Fedora, with and without vcpkg, tested from absolute fresh in VMs (Fedora 40 and Ubuntu 22.04 LTS).

Should be ready to merge!

Jayman2000 commented on README.md:
```diff
@@ -92,17 +92,24 @@ The milestone needs testing on all platforms. Please report issues when found.
-#### Building - Windows
-1. Make sure that you have Git and Visual Studio 2022 with the “Desktop development with C++” workload and the “C++ MFC for latest v143 build tools (x86 & x64)” component. If you don’t already have those installed or you aren’t sure, then open an elevated Command Prompt and run:
+### Dependencies
+The build process uses [**CMake**](https://cmake.org/) and, by default, [**Ninja**](https://ninja-build.org/). You must install these; the project cannot locate them for you. The source code depends on [**SDL2**](https://github.com/libsdl-org/SDL) and [**zlib**](https://github.com/madler/zlib). You can supply these dependencies yourself via your system's library management tools, or the build sys ...
```
The SDL2 link goes to the SDL repo’s branch for SDL3. I would change it to this:

```suggestion
The build process uses [**CMake**](https://cmake.org/) and, by default, [**Ninja**](https://ninja-build.org/). You must install these; the project cannot locate them for you. The source code depends on [**SDL2**](https://github.com/libsdl-org/SDL/tree/SDL2) and [**zlib**](https://github.com/madler/zlib). You can supply these dependencies yourself via your system's library management tools, or the build system can locate the dependencies for you using [vcpkg](https://github.com/microsoft/vcpkg), a cross-platform dependency-management system developed by Microsoft. The official builds source their dependencies from vcpkg.
```

Jayman2000 commented on README.md:
```diff
@@ -92,17 +92,24 @@ The milestone needs testing on all platforms. Please report issues when found.
+The build process uses [**CMake**](https://cmake.org/) and, by default, [**Ninja**](https://ninja-build.org/). You must install these; the project cannot locate them for you. The source code depends on [**SDL2**](https://github.com/libsdl-org/SDL) and [**zlib**](https://github.com/madler/zlib). You can supply these dependencies yourself via your system's library management tools, or the build sys ...
 
+### Installing and using vcpkg
+* When building for Windows, vcpkg is already installed and configured when using the Developer Tools Command Prompt.
```
The Command Prompt that we tell users to use doesn’t have the phrase “Developer Tools” in its title. I would reword that sentence to make it more clear that the vcpkg thing applies to all of the “-for VS 2022” Command Prompts:

```suggestion
* When building for Windows, vcpkg is already installed and configured when using any of the Visual Studio Command Prompts.
```

Jayman2000 commented on README.md:
```diff
@@ -114,61 +121,144 @@ The milestone needs testing on all platforms. Please report issues when found.
 
-2. Open a “x64 Native Tools Command Prompt” and run:
+    **NOTE:**
+    Builds _must_ be performed in the “x64 Native Tools Command Prompt” (or PowerShell) provided by Visual Studio, _not_ a standard Command Prompt (or PowerShell). The Native Tools CLIs will already have vcpkg installed and configured for use, so no dependency management is needed.
```
There is no “x64 Native Tools PowerShell”. I would remove that first reference to PowerShell:

```suggestion
    Builds _must_ be performed in the “x64 Native Tools Command Prompt” provided by Visual Studio, _not_ a standard Command Prompt (or PowerShell). The Native Tools CLIs will already have vcpkg installed and configured for use, so no dependency management is needed.
```

Jayman2000 commented on README.md:
````diff
@@ -114,61 +121,144 @@ The milestone needs testing on all platforms. Please report issues when found.
+    * APT users (Debian, Ubuntu)
+        ```sh
+        sudo apt update
+        sudo apt install -y --no-install-recommends git ninja-build cmake g++ libncurses-dev
+        ```
+    * DNF users (Red Hat, Fedora)
+        ```sh
+        sudo dnf update --refresh
+        sudo dnf install -y git ninja-build cmake gcc-c++ ncurses-devel
````
Why do we need ncurses?

Lgt2x replied:
#407 removes every use of ncurses

tophyr replied:
ncurses provides `term.h`, which [we do still seem to need](https://github.com/DescentDevelopers/Descent3/blob/main/linux/lnxapp.cpp#L75)

winterheart replied:
Seems this is oversight by me. `term.h` is unused and can be safely removed from code.

Jayman2000 commented on README.md:
```diff
@@ -114,61 +121,144 @@ The milestone needs testing on all platforms. Please report issues when found.
+2. **Acquire the library dependencies.**
+
+    * If you would like to use vcpkg:
+        1. Clone vcpkg
```
This line is missing a colon:

```suggestion
        1. Clone vcpkg:
```

Jayman2000 reviewed (changes requested):
I tested this PR and got it to work in these situations:

- Windows 10 x86_64

- Ubuntu 24.04 x86_64 without vcpkgs

- Fedora 40 x86_64 without vcpkgs.

I was not able to get it to work in these situations:

- Ubuntu 24.04 x86_64 with vcpkgs. The game compiles fine, but when I run it, I get this error:

    ```
    SDL: SDL_Init() failed! rc == (-1).
    SDL_GetError() reports "dsp: No such audio device".
    ```

    [Here’s a log of me building and running the game.](https://gist.github.com/Jayman2000/c88b5fb02bf69644d8ac8bf7e7b0d95d)

- Fedora 40 x86_64 with vcpkgs. When I run `cmake --preset linux`, it fails with several error messages. The most important one seems to be:

    ```
    error: building libsystemd:x64-linux failed with: BUILD_FAILED
    ```

    [Here’s a log of me trying to build the game.](https://gist.github.com/Jayman2000/b4eca3a1038d95d01f45bff504e3beb8)

tophyr commented:
Incorporated @Jayman2000 README feedback
Fixed audio runtime errors on Linux by adding ALSA and PulseAudio packages to requirements list
Fixed SystemD build error by updating vcpkg hash
Removed `term.h` reference and dropped ncurses dependency
Squashed commit stack

tophyr commented:
One thing worth noting is that the mouse behavior is wonky, in both of my Linux builds. But, I only have Linux VMs - I'd appreciate anyone with a bare-metal Linux machine testing this and running the binary and checking if the mouse is usable (both in the GUI and in the game).

Lgt2x commented on CMakeLists.txt:
```diff
@@ -5,21 +5,41 @@ if(NOT CMAKE_BUILD_TYPE AND NOT DEFINED ENV{CMAKE_BUILD_TYPE})
   set(CMAKE_BUILD_TYPE "Debug" CACHE STRING "default build type")
 endif()
 
-project(Descent3
```
this should still be on top

Lgt2x commented:
> One thing worth noting is that the mouse behavior is wonky, in both of my Linux builds. But, I only have Linux VMs - I'd appreciate anyone with a bare-metal Linux machine testing this and running the binary and checking if the mouse is usable (both in the GUI and in the game).

I don't have such a problem on my machine (not a VM)

The code looks mostly good to me now, very good job. After the last comment is addressed we're all set. I'll let @Jayman2000 approve additionally

Jayman2000 commented on README.md:
```diff
@@ -114,61 +121,144 @@ The milestone needs testing on all platforms. Please report issues when found.
+
+2. **Clone the Descent3 source code.**
 
+    Open a Developer Command Prompt (or Developer PowerShell, x64 Native Tools Command Prompt, etc if you want a specific environment) and run:
```
Why did we switch from telling people to use an x64 Native Tools Command Prompt to telling people to use any of the “-for VS 2022” command prompts?

tophyr replied:
Hubris! I wanted to make sure it was clear that PowerShell would work as well, and overreached. I've switched the README back to saying
> Open a "x64 Native Tools Command Prompt" or "x64 Native Tools PowerShell" and run:

Jayman2000 replied:
I don’t have an x64 Native Tool PowerShell shortcut in my Start menu:

![My Start menu has an entry named “Visual Studio 2022”. Beneath that entry, there’s a folder that’s also named “Visual Studio 2022”. In that folder there are several entries: “Debuggable Package Manager”, “Developer Command Prompt for VS 2022”, “Developer PowerShell for VS 2022”, “x64 Native Tools Command Prompt for VS 2022”, “x64_x86 Cross Tools Command Prompt for VS 2022”, “x86 Native Tools Command Prompt for VS 2022” and “x86_64 Cross Tools Command Prompt for VS 2022”. Beneath all of the entries in the “Visual Studio 2022” folder is an entry named “Visual Studio Installer”.](https://github.com/user-attachments/assets/40e4b041-a56c-4948-b46f-1cd321442aba)

How do I open an x64 Native Tools PowerShell? Do I need to do something in order to get a Start menu shortcut for it?

Jayman2000 reviewed (changes requested):
I tested this PR and got it to work in these situations:

- Windows 10 x86_64 using an “x64 Native Tools Command Prompt for VS 2022”

- Ubuntu 24.04 x86_64 with vcpkg.

- Ubuntu 24.04 x86_64 without vcpkg.

- Fedora 40 x86_64 with vcpkg.

- Fedora 40 x86_64 without vcpkg.

I was not able to get it to work in these situations:

- Windows 10 x86_64 using a “Developer Command Prompt for VS 2022”. When I try to run `cmake --preset win`, it fails with this error:

    ```
      Could not find a configuration file for package "SDL2" that is compatible
      with requested version "".
    ```
    
    [Here’s a log of me trying to build the game.](https://gist.github.com/Jayman2000/e37befae1ff73cd128c7fde8d2b74218)

> Incorporated @Jayman2000 README feedback

I noticed that the README still mentions Android. Do you think that it would be better if the README mentioned Android even though we don’t support it yet?

> One thing worth noting is that the mouse behavior is wonky, in both of my Linux builds. But, I only have Linux VMs - I'd appreciate anyone with a bare-metal Linux machine testing this and running the binary and checking if the mouse is usable (both in the GUI and in the game).

I get the same problem when I run the `main` branch in a VM, so I don’t think that it’s related to the changes in this PR.

tophyr commented:
Updated, hopefully for the final time, to address @Jayman2000's Developer Command Prompt for VS 2022 failure - README now excludes that configuration, like it probably should've from the start.

Also added SDL2, gtest and zlib back to Brewfile, to support non-vcpkg building there as well.

Lgt2x reviewed (approved):
Thanks for completing this :)
